File: core-v3/src/main/java/com/adyen/checkout/core/api/BaseHttpUrlConnectionFactory.java
/*
 * Copyright (c) 2018 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by timon on 24/04/2018.
 */

package com.adyen.checkout.core.api;

import static com.adyen.checkout.core.api.SSLSocketUtil.TLS_SOCKET_FACTORY;

import androidx.annotation.NonNull;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;

import javax.net.ssl.HttpsURLConnection;

abstract class BaseHttpUrlConnectionFactory {

    static {
        HttpsURLConnection.setDefaultSSLSocketFactory(TLS_SOCKET_FACTORY);
    }

    @NonNull
    HttpURLConnection createHttpUrlConnection(@NonNull String url) throws IOException {
        final HttpURLConnection urlConnection = (HttpURLConnection) new URL(url).openConnection();

        if (urlConnection instanceof HttpsURLConnection) {
            ((HttpsURLConnection) urlConnection).setSSLSocketFactory(TLS_SOCKET_FACTORY);
            return urlConnection;
        } else {
            return handleInsecureConnection(urlConnection);
        }
    }

    @NonNull
    abstract HttpURLConnection handleInsecureConnection(@NonNull HttpURLConnection httpUrlConnection);
}

File: checkout-core/src/main/java/com/adyen/checkout/core/api/BaseHttpUrlConnectionFactory.kt
/*
 * Copyright (c) 2020 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 17/12/2020.
 */
package com.adyen.checkout.core.api

import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import javax.net.ssl.HttpsURLConnection

internal abstract class BaseHttpUrlConnectionFactory {

    @Throws(IOException::class)
    fun createHttpUrlConnection(url: String): HttpURLConnection {
        val urlConnection = URL(url).openConnection() as HttpURLConnection
        return if (urlConnection is HttpsURLConnection) {
            urlConnection.sslSocketFactory = SSLSocketUtil.TLS_SOCKET_FACTORY
            urlConnection
        } else {
            handleInsecureConnection(urlConnection)
        }
    }

    abstract fun handleInsecureConnection(httpUrlConnection: HttpURLConnection): HttpURLConnection

    companion object {
        init {
            HttpsURLConnection.setDefaultSSLSocketFactory(SSLSocketUtil.TLS_SOCKET_FACTORY)
        }
    }
}
